### PR TITLE
Add sphinx doc comments regarding exceptions in the optimize method

### PR DESCRIPTION
**Pull request by yuk1ty**
## Motivation

See issue #1618 for the original motivation. This issue is good for me to look through Optuna.

## Description of the changes

I'm trying to work on the issue #1618. This PR is a partial fulfillment of adding `Raises:` comments to docs. To begin with, I added a comment to `optimize` method.

Could anyone tell me about the following questions?
- How much comment should I write, for instance, if additional exceptions are thrown from inside of `optimize` method? 
- Is the comment in my PR sufficient?

### CODE REVIEW

File: optuna/study.py [hunk 1]
```diff
@@ -304,6 +304,12 @@ def objective(trial):
                 Flag to show progress bars or not. To disable progress bar, set this ``False``.
                 Currently, progress bar is experimental feature and disabled
                 when ``n_jobs`` :math:`\\ne 1`.
+        
+        Raises:
+            TypeError:
+                If the catch argument is not a tuple.
```
Comment (crcrpar) [inline optuna/study.py:310]: [nit-picking]
```suggestion
                If the ``catch`` argument is not a tuple.
```

Comment (yuk1ty) [reply optuna/study.py]: Ah, thank you! I fixed it.

File: optuna/study.py [hunk 2]
```diff
@@ -304,6 +304,12 @@ def objective(trial):
                 Flag to show progress bars or not. To disable progress bar, set this ``False``.
                 Currently, progress bar is experimental feature and disabled
                 when ``n_jobs`` :math:`\\ne 1`.
+
```
Comment (HideakiImamura) [inline optuna/study.py:307]: You can pass the CI by removing white spaces.
```suggestion

```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I have a minor comment.

File: optuna/study.py (hunk 2 above)
Comment (yuk1ty) [reply optuna/study.py]: I missed it. Thank you! Fixed.

**Comment by hvy** [comment]
Thanks a lot for the PR, the changed looks good overall. I'm wondering though if we should skip documenting the `TypeError` though.

Python is a dynamically typed language and while we do check the types in certain APIs, we also don't in many (nor cover in unit tests). The expected types are communicated via the type hints and we sort of trust the user to follow these types. Unfortunately, the behavior is not really consistent and maintaining the documentation might not be worth it, at least for the time being.

**Comment by yuk1ty** [comment]
> The expected types are communicated via the type hints and we sort of trust the user to follow these types. Unfortunately, the behavior is not really consistent and maintaining the documentation might not be worth it, at least for the time being.

I understood & I agree with that point so let me remove the `TypeError` comment.

**Review by HideakiImamura — APPROVED** [review]
LGTM!

File: optuna/study.py [hunk 3]
```diff
@@ -304,6 +304,10 @@ def objective(trial):
...
+        Raises:
+            RuntimeError:
+                If nested invocation of this method occurred.
```
Comment (hvy) [inline optuna/study.py:310]: Sorry for the late review but could you just fix the tense to be consistent with e.g. https://github.com/optuna/optuna/blob/master/optuna/storages/_base.py#L105 ?
```suggestion
                If nested invocation of this method occurs.
```

Comment (yuk1ty) [reply optuna/study.py:310]: Sure!

**Review by hvy — APPROVED** [review]
Thanks for the swift response, LGTM!

**Comment by hvy** [comment]
As for how to document exceptions in general, I think it's difficult to say. If we want to make it a part of the interface, then it should be documented. I guess we have to deal with this case by case since documenting every possible exception is infeasible.

**Pull request merged**